contracts/KidnapInsurance.sol

```diff
@@ -3,79 +3,89 @@ pragma solidity ^0.8.0;
 
 contract KidnapInsurance {
 
-    event Abducted();
+    event Abducted(address indexed kidnapper, uint256 requestedRansom);
     event Rescued();
     event FuckYou();
 
-    address public kidnapper;
-    address public friends;
-    address payable public ransomReceiver;
+    address payable public kidnapper;
+    address payable public friends;
+    uint256 public ransomAmount;
 
     uint256 public kidnappedTime;
-    bool public vetoed;
+    uint256 constant delay = 3 days;
 
-    uint256 constant ransom = 1000e18; // 1k ETH
-    uint256 constant veto = 2000e18; // 2k ETH
-    uint256 constant delay = 3 days; // 3 days
-
-    modifier onlyKidnapper() {
-        require(msg.sender == kidnapper, "Kidnapper: only kidnapper");
-        _;
+    constructor(address payable _friends) payable {
+        friends = _friends;
     }
 
-    modifier onlyFriends() {
-        require(msg.sender == friends, "Kidnapper: only friends");
-        _;
-    }
+    // Hash of the password to initiate ransom withdraw
+    bytes32 constant WITHDRAW_HASH = 0x81387e6fe7c09e310810396aee0057e20224fd16320359fae10f45392cba8c80;
 
-    constructor(address _kidnapper, address _friends) payable {
-        require(msg.value == ransom, "Kidnapper: Deposit must be ransom");
-        kidnapper = _kidnapper;
-        friends = _friends;
+    // Generate a hash to show the sender knows the password. Can also be done off-chain.
+    function generateHash(string memory _password) public view returns (bytes32 hash) {
+        bytes32 passwordHash = keccak256(abi.encodePacked(_password));
+        hash = keccak256(abi.encodePacked(msg.sender, passwordHash));
     }
 
-    // Kidnapper Funcs
-    function initiateRansomWithdraw(address payable _receiver) onlyKidnapper public {
-        // Must be first kidnapp
-        require(kidnappedTime == 0, "Kidnapper: Already kindapped");
+    function initiateRansomWithdraw(bytes32 _hash, uint256 _ransomAmount) public {
+        require(kidnapper == address(0), "Kidnapper: Already kindapped");
+        require(address(this).balance >= _ransomAmount, "Kidnapper: Not enough ransom available");
+
+        // Check if the kidnapper knows the password, front running proof
+        bytes32 correctHash = keccak256(abi.encodePacked(msg.sender, WITHDRAW_HASH));
+        require(correctHash == _hash, "Kidnapper: Invalid password or sender");
```

Comment:
How is this hash check secure? Given that the `WITHDRAW_HASH` is publicly available, anyone can simply compute `correctHash` off-chain and pass it as the `_hash` to `initiateRansomWithdraw` and thus initiating the withdraw process. 

I don't see how you can check whether the kidnappers actually know the password without passing it via calldata here, which would be front-runnable again. 

That's why I chose the private key route and limited the initiateRandsomWithdraw func to `onlyKidnapper`.

Reply:
You are right! Give me a minute, I'll propose a different approach with a password :)

Reply:
@hilmarx I think this should work -> 676b78d8e5dc0050e99d7c677a752ed7c0920da2